File: problems/0008_string-to-integer-atoi/one-loop/solution.go
func myAtoi(str string) int {
    sLen := len(str)
    if sLen == 0 {
        return 0
    }
    i := 0
    for i < sLen && str[i] == ' ' {
        i++
    }
    sign := 1
    if i < sLen {
        if str[i] == '-' {
            i++
            sign = -1
        } else if str[i] == '+' {
            i++
        }
    }
    result := 0
    for i < sLen && '0' <= str[i] && str[i] <= '9' {
        result = result * 10 + sign * int(str[i] - '0')
        if result < MinInt32 {
            return MinInt32
        } else if result > MaxInt32 {
            return MaxInt32
        }
        i++
    }
    return result
}

const MaxInt32 = int(^uint32(0) >> 1)
const MinInt32 = -MaxInt32 - 1

File: problems/0008_string-to-integer-atoi/one-loop/Solution.rs
use std::i32;

impl Solution {
    pub fn my_atoi(str: String) -> i32 {
        let s_len = str.len();
        if s_len == 0 {
            return 0;
        }
        let (mut i, ptr) = (0, str.as_ptr());
        while i < s_len && unsafe { *(ptr.add(i)) } as char == ' ' {
            i += 1;
        }
        let mut sign = 1i64;
        if i < s_len {
            let c = unsafe { *(ptr.add(i)) } as char;
            if c == '-' {
                i += 1;
                sign = -1i64;
            } else if c == '+' {
                i += 1;
            }
        }
        let mut result = 0i64;
        while i < s_len {
            let c = unsafe { *(ptr.add(i)) } as char;
            if c < '0' || '9' < c {
                break;
            }
            result = result * 10 + sign * (c as i64 - '0' as i64);
            if result < i32::MIN as i64 {
                return i32::MIN;
            } else if result > i32::MAX as i64 {
                return i32::MAX;
            }
            i += 1;
        }
        result as i32
    }
}
